src/save-restore.c

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright Red Hat
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include <gio/gio.h>
+
+#include "xdp-context.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-portal-config.h"
+#include "xdp-session.h"
+#include "xdp-utils.h"
+
+#include "save-restore.h"
+
+struct _XdpSaveRestore
+{
+  XdpDbusSaveRestoreSkeleton parent_instance;
+
+  XdpDbusImplSaveRestore *impl;
+  GCancellable *cancellable;
+};
+
+#define XDP_TYPE_SAVE_RESTORE (xdp_save_restore_get_type ())
+G_DECLARE_FINAL_TYPE (XdpSaveRestore,
+                      xdp_save_restore,
+                      XDP, SAVE_RESTORE,
+                      XdpDbusSaveRestoreSkeleton)
+
+static void xdp_save_restore_iface_init (XdpDbusSaveRestoreIface *iface);
+
+G_DEFINE_FINAL_TYPE_WITH_CODE (XdpSaveRestore,
+                               xdp_save_restore,
+                               XDP_DBUS_TYPE_SAVE_RESTORE_SKELETON,
+                               G_IMPLEMENT_INTERFACE (XDP_DBUS_TYPE_SAVE_RESTORE,
+                                                      xdp_save_restore_iface_init));
+
+struct _XdpSaveRestoreSession
+{
+  XdpSession parent;
+
+  gboolean closed;
+};
+
+#define XDP_TYPE_SAVE_RESTORE_SESSION (xdp_save_restore_session_get_type ())
+G_DECLARE_FINAL_TYPE (XdpSaveRestoreSession,
+                      xdp_save_restore_session,
+                      XDP, SAVE_RESTORE_SESSION,
+                      XdpSession)
+
+G_DEFINE_TYPE (XdpSaveRestoreSession,
+               xdp_save_restore_session,
+               xdp_session_get_type ())
+
+static void
+xdp_save_restore_session_close (XdpSession *session)
+{
+  XdpSaveRestoreSession *save_restore_session = XDP_SAVE_RESTORE_SESSION (session);
+
+  save_restore_session->closed = TRUE;
```

Comment:
I think a couple of the session implementations do nothing but keep track of this closed variable.

Maybe that should move up into the `XdpSession` object?

On the other hand, this class separation actually lets us check that the caller is giving us a session path to a session of a type we expect. So we'd need to have a subclass anyway. Or at the very least some sort of accounting of what portal created any given session